# HG changeset patch
# User joe
# Date 1541371233 25200
#      Sun Nov 04 15:40:33 2018 -0700
# Node ID efa5461222f86faba6fd755f012155f2c823b862
# Parent  ecd46ac895975d8ddd30c1caf804053c2a5e8637
rio keybindings to send or plumb the current line or selection

<kvik> joe9: here's a rio patch. hope it applies, i edited it by hand. http://okturing.com/src/5315/body  [19:46]
<kvik> joe9: ^Q will "send" a current line or selection. you'll notice there's also ^P which does "plumb"  [19:48]
<mischief> do you hate the mouse  [19:51]
<joe9> kvik: Thanks.  [19:54]
<kvik> mischief: no, just the stupid "find and double click the start or end of line, do the menu dance" for a really common task  [20:00]

diff -r ecd46ac89597 -r efa5461222f8 sys/src/cmd/rio/dat.h
--- a/sys/src/cmd/rio/dat.h	Mon Oct 22 10:35:59 2018 -0700
+++ b/sys/src/cmd/rio/dat.h	Sun Nov 04 15:40:33 2018 -0700
@@ -210,6 +210,7 @@
 void		wscrdraw(Window*);
 void		wscroll(Window*, int);
 void		wselect(Window*);
+void		wsend(Window*);
 void		wsendctlmesg(Window*, int, Rectangle, void*);
 void		wsetcursor(Window*, int);
 void		wsetname(Window*);
diff -r ecd46ac89597 -r efa5461222f8 sys/src/cmd/rio/rio.c
--- a/sys/src/cmd/rio/rio.c	Mon Oct 22 10:35:59 2018 -0700
+++ b/sys/src/cmd/rio/rio.c	Sun Nov 04 15:40:33 2018 -0700
@@ -809,21 +809,7 @@
 		break;
 
 	case Send:
-		getsnarf();
-		wsnarf(w);
-		if(nsnarf == 0)
-			break;
-		if(w->rawing){
-			waddraw(w, snarf, nsnarf);
-			if(snarf[nsnarf-1]!='\n' && snarf[nsnarf-1]!='\004')
-				waddraw(w, L"\n", 1);
-		}else{
-			winsert(w, snarf, nsnarf, w->nr);
-			if(snarf[nsnarf-1]!='\n' && snarf[nsnarf-1]!='\004')
-				winsert(w, L"\n", 1, w->nr);
-		}
-		wsetselect(w, w->nr, w->nr);
-		wshow(w, w->nr);
+		wsend(w);
 		break;
 
 	case Scroll:
diff -r ecd46ac89597 -r efa5461222f8 sys/src/cmd/rio/wind.c
--- a/sys/src/cmd/rio/wind.c	Mon Oct 22 10:35:59 2018 -0700
+++ b/sys/src/cmd/rio/wind.c	Sun Nov 04 15:40:33 2018 -0700
@@ -677,6 +677,24 @@
 			wsetselect(w, w->qh, w->qh);
 			wshow(w, w->q0);
 			return;
+		case 0x10:	/* ^P: plumb */
+			wplumb(w);
+			return;
+		case 0x11:	/* ^Q: send line or selection */
+			if(w->q0 != w->q1){
+				wsend(w);
+				return;
+			}
+			if(w->q0==0 || w->q0==w->qh || w->r[w->q0-1]=='\n')
+				q0 = w->q0;
+			else
+				q0 = w->q0 - wbswidth(w, 0x15);
+			q1 = w->q0;
+			while(q1 < w->nr && w->r[q1]!='\n')
+				q1++;
+			wsetselect(w, q0, q1);
+			wsend(w);
+			return;
 		}
 	if(w->rawing && (w->q0==w->nr || w->mouseopen)){
 		waddraw(w, &r, 1);
@@ -1788,3 +1806,23 @@
 {
 	return runetobyte(w->r, w->nr, ip);
 }
+
+void
+wsend(Window *w)
+{
+	getsnarf();
+	wsnarf(w);
+	if(nsnarf == 0)
+		return;
+	if(w->rawing){
+		waddraw(w, snarf, nsnarf);
+		if(snarf[nsnarf-1]!='\n' && snarf[nsnarf-1]!='\004')
+			waddraw(w, L"\n", 1);
+	}else{
+		winsert(w, snarf, nsnarf, w->nr);
+		if(snarf[nsnarf-1]!='\n' && snarf[nsnarf-1]!='\004')
+			winsert(w, L"\n", 1, w->nr);
+	}
+	wsetselect(w, w->nr, w->nr);
+	wshow(w, w->nr);
+}
